lib/experimental/text_layouter.nim

```diff
@@ -0,0 +1,1762 @@
+
+func retreat(self: var LytSolution) =
+  ## Move back a knot.
+  self.index -= 1
+
+func curKnot(self: LytSolution): int =
+  ## The currently indexed knot.
+  return self.knots[self.index]
+
+func curSpan(self: LytSolution): int =
+  return self.spans[self.index]
+
+func curIntercept(self: LytSolution): float =
+  return self.intercepts[self.index]
+
+func curGradient(self: LytSolution): float =
+  return self.gradients[self.index]
+
+func curLayout(self: LytSolution): Layout = self.layouts[self.index]
+func curIndex(self: LytSolution): int = self.index
+
+func curValueAt(self: LytSolution, margin: int): float =
+  ## The value (cost) extrapolated for margin m from the current knot.
+  # Since a LytSolution's cost is represented by a piecewise linear function,
+  # the extrapolation in this case is linear, from the current knot.
+  return self.curIntercept() + self.curGradient() * float(
+    margin - self.curKnot())
+
+func nextKnot(self: LytSolution): int =
+  ## The knot after the once currently indexed.
+  if self.index + 1 >= self.knots.len:
+    infty
+  else:
+    self.knots[self.index + 1]
+
+proc moveToMargin(self: var LytSolution, margin: int) =
+  ## Adjust the index so m falls between the current knot and the next.
+  if self.curKnot() > margin:
+    while self.curKnot() > margin:
+      self.retreat()
+  else:
+    while self.nextKnot() <= margin and self.nextKnot() != infty:
+      self.advance()
+      # info "Advancing to position", self.curIndex(),
+      #   "next knot is", self.nextKnot(), "margin is", margin,
+      #   self.nextKnot() <= margin
+
+
+#==========================  LytSolution factory  ===========================#
+
+proc add(
+    self: var LytSolution,
+    knot, span: int,
+    intercept, gradient: float,
+    layout: Layout
+  ) =
+  ## Add a segment to a LytSolution under construction.
+  ##
+  ## The function performs basic consistency checks, and eliminates
+  ## redundant segments that are linear extrapolations of those that
+  ## precede them.
+
+  if self.isNil():
+    new(self)
+
+  if 0 < self.knots.len:
+    # Don't add a knot if the new segment is a linear extrapolation of
+    # the last.
+    let
+      kLast = self.knots[^1]
+      sLast = self.spans[^1]
+      iLast = self.intercepts[^1]
+      gLast = self.gradients[^1]
+
+    if (span == sLast and gradient == gLast and
+        iLast + (knot - kLast) * gLast == intercept):
+      return
+
+  if knot < 0 or span < 0 or intercept < 0 or gradient < 0:
+    raiseAssert(
+      "Internal error: bad layout: " &
+        &"(k {knot}, s {span}, i {intercept}, g {gradient})")
+
+  self.knots.add knot
+  self.spans.add span
+  self.intercepts.add intercept
+  self.gradients.add gradient
+  self.layouts.add layout
+
+
+#=====================  LytSolution manipulation logic  =====================#
+
+
+proc minSolution(solutions: seq[LytSolution]): Option[LytSolution] =
+  ## Form the piecewise minimum of a sequence of LytSolutions.
+
```

Comment:
observation:

I find these visual breaks have me jumping back up to ensure I suddenly didn't lose context -- destroys my skim speed.

It's not a problem when I'm reading carefully, but quickly skimming is really hard because I'm purposefully not reading everything carefully, so I suddenly doubt things and wonder if I missed something and now I've fallen out of skim mode and am back tracking.

Reply:
I don't understand what "visual breaks" do you refer to. Do you mean the `#=====================  LytSolution manipulation logic  =====================#` comment above the line you commented on?

Reply:
```nim
  ## Form the piecewise minimum of a sequence of LytSolutions.
  # (blank line is the visual break)
  ## more docs
```

The blank line in the block doc comment, I left it as an observation.

It's merely a data point so over some period of time (bigger than this PR) we eventually figure out a style guide.

Reply:
fixed to continuous comments now